### ISSUE TITLE
I get an error after running "npm run server:ssr"

### ISSUE THREAD

**Issue by khaled6mmd** [angular/universal#1576]
after done all the steps to convert to angular universal 
from [https://angular.io/guide/universal](url)
(I noticed also that there is no "webpack.server.config.js" file anymore in created files)
got the following 

error msg in the terminal 
```
 coupons@0.0.0 serve:ssr C:\Users\Khaled\Desktop\coupons
 node dist/coupons/server/main.js

C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1
!function(e,a){for(var i in a)e[i]=a[i]}(exports,function(modules){var installedModules={};function __webpack_require__(moduleId){if(installedModules[moduleId])return installedModules[moduleId].exports;var module=installedModules[moduleId]={i:moduleId,l:!1,exports:{}};return modules[moduleId].call(module.exports,module,module.exports,__webpack_require__),module.l=!0,module.exports}return __webpack_require__.m=modules,__webpack_require__.c=installedModules,__webpack_require__.d=function(exports,name,getter){__webpack_require__.o(exports,name)||Object.defineProperty(exports,name,{enumerable:!0,get:getter})},__webpack_require__.r=function(exports){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(exports,"__esModule",{value:!0})},__webpack_require__.t=function(value,mode){if(1&mode&&(value=__webpack_require__(value)),8&mode)return value;if(4&mode&&"object"==typeof

ReferenceError: Event is not defined
    at Module.bhfF (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:2114998)
    at __webpack_require__ (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:295)
    at Object.ZAI4 (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:2025599)
    at __webpack_require__ (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:295)
    at Object.24aS (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:694036)    at __webpack_require__ (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:295)
    at Object.K011 (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:1480280)
    at __webpack_require__ (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:295)
    at Object.uj+Y (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:3894777)
    at __webpack_require__ (C:\Users\Khaled\Desktop\coupons\dist\coupons\server\main.js:1:295)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! coupons@0.0.0 serve:ssr: `node dist/coupons/server/main.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the coupons@0.0.0 serve:ssr script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Khaled\AppData\Roaming\npm-cache\_logs\2020-03-27T03_43_54_076Z-debug.log
```
the log of the issue
```
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'serve:ssr'
1 verbose cli ]
2 info using npm@6.12.0
3 info using node@v12.13.0
4 verbose run-script [ 'preserve:ssr', 'serve:ssr', 'postserve:ssr' ]
5 info lifecycle coupons@0.0.0~preserve:ssr: coupons@0.0.0
6 info lifecycle coupons@0.0.0~serve:ssr: coupons@0.0.0
7 verbose lifecycle coupons@0.0.0~serve:ssr: unsafe-perm in lifecycle true
8 verbose lifecycle coupons@0.0.0~serve:ssr: PATH: C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin;C:\Users\Khaled\Desktop\coupons\node_modules\.bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files (x86)\Brackets\command;C:\Program Files\nodejs\;C:\ProgramData\chocolatey\bin;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Users\Khaled\AppData\Local\Microsoft\WindowsApps;C:\Users\Khaled\AppData\Roaming\npm;C:\Users\Khaled\AppData\Local\GitHubDesktop\bin;C:\Users\Khaled\AppData\Local\Programs\Microsoft VS Code\bin;C:\Program Files\heroku\bin
9 verbose lifecycle coupons@0.0.0~serve:ssr: CWD: C:\Users\Khaled\Desktop\coupons
10 silly lifecycle coupons@0.0.0~serve:ssr: Args: [ '/d /s /c', 'node dist/coupons/server/main.js' ]
11 silly lifecycle coupons@0.0.0~serve:ssr: Returned: code: 1  signal: null
12 info lifecycle coupons@0.0.0~serve:ssr: Failed to exec serve:ssr script
13 verbose stack Error: coupons@0.0.0 serve:ssr: `node dist/coupons/server/main.js`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:210:5)
13 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:210:5)
13 verbose stack     at maybeClose (internal/child_process.js:1021:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
14 verbose pkgid coupons@0.0.0
15 verbose cwd C:\Users\Khaled\Desktop\coupons
16 verbose Windows_NT 10.0.18363
17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "serve:ssr"
18 verbose node v12.13.0
19 verbose npm  v6.12.0
20 error code ELIFECYCLE
21 error errno 1
22 error coupons@0.0.0 serve:ssr: `node dist/coupons/server/main.js`
22 error Exit status 1
23 error Failed at the coupons@0.0.0 serve:ssr script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]

```

package.json

```
{
  "name": "coupons",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e",
    "dev:ssr": "ng run coupons:serve-ssr",
    "serve:ssr": "node dist/coupons/server/main.js",
    "build:ssr": "ng build --prod && ng run coupons:server:production",
    "prerender": "ng run coupons:prerender"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "~9.0.5",
    "@angular/common": "~9.0.5",
    "@angular/compiler": "~9.0.5",
    "@angular/core": "~9.0.5",
    "@angular/fire": "^5.4.2",
    "@angular/forms": "~9.0.5",
    "@angular/platform-browser": "~9.0.5",
    "@angular/platform-browser-dynamic": "~9.0.5",
    "@angular/platform-server": "~9.0.5",
    "@angular/router": "~9.0.5",
    "@nguniversal/express-engine": "^9.0.2",
    "express": "^4.15.2",
    "firebase": "^7.11.0",
    "ng2-tooltip-directive": "^2.9.19",
    "ngx-clipboard": "^13.0.0",
    "ngx-owl-carousel-o": "^2.0.3",
    "rxjs": "~6.5.4",
    "tslib": "^1.10.0",
    "uuid": "^7.0.2",
    "zone.js": "~0.10.2"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.900.5",
    "@angular/cli": "~9.0.5",
    "@angular/compiler-cli": "~9.0.5",
    "@angular/language-service": "~9.0.5",
    "@nguniversal/builders": "^9.0.2",
    "@types/express": "^4.17.0",
    "@types/jasmine": "~3.3.8",
    "@types/jasminewd2": "~2.0.3",
    "@types/node": "^12.11.1",
    "autoprefixer": "^9.7.4",
    "codelyzer": "^5.1.2",
    "jasmine-core": "~3.4.0",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~4.1.0",
    "karma-chrome-launcher": "~2.2.0",
    "karma-coverage-istanbul-reporter": "~2.0.1",
    "karma-jasmine": "~2.0.1",
    "karma-jasmine-html-reporter": "^1.4.0",
    "postcss": "^7.0.27",
    "postcss-cli": "^7.1.0",
    "protractor": "~5.4.0",
    "ts-node": "~7.0.0",
    "tslint": "~5.15.0",
    "typescript": "~3.7.5"
  },
  "prettier": {
    "singleQuote": true,
    "printWidth": 100,
    "semi": true,
    "bracketSpacing": true,
    "arrowParens": "always"
  }
}

```
server.ts

```
import 'zone.js/dist/zone-node';

import { ngExpressEngine } from '@nguniversal/express-engine';
import * as express from 'express';
import { join } from 'path';

import { AppServerModule } from './src/main.server';
import { APP_BASE_HREF } from '@angular/common';
import { existsSync } from 'fs';

// The Express app is exported so that it can be used by serverless Functions.
export function app() {
  const server = express();
  const distFolder = join(process.cwd(), 'dist/coupons/browser');
  const indexHtml = existsSync(join(distFolder, 'index.original.html')) ? 'index.original.html' : 'index';

  // Our Universal express-engine (found @ https://github.com/angular/universal/tree/master/modules/express-engine)
  server.engine('html', ngExpressEngine({
    bootstrap: AppServerModule,
  }));

  server.set('view engine', 'html');
  server.set('views', distFolder);

  // Example Express Rest API endpoints
  // app.get('/api/**', (req, res) => { });
  // Serve static files from /browser
  server.get('*.*', express.static(distFolder, {
    maxAge: '1y'
  }));

  // All regular routes use the Universal engine
  server.get('*', (req, res) => {
    res.render(indexHtml, { req, providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }] });
  });

  return server;
}

function run() {
  const port = process.env.PORT || 4000;

  // Start up the Node server
  const server = app();
  server.listen(port, () => {
    console.log(`Node Express server listening on http://localhost:${port}`);
  });
}

// Webpack will replace 'require' with '__webpack_require__'
// '__non_webpack_require__' is a proxy to Node 'require'
// The below code is to ensure that the server is run only when not requiring the bundle.
declare const __non_webpack_require__: NodeRequire;
const mainModule = __non_webpack_require__.main;
const moduleFilename = mainModule && mainModule.filename || '';
if (moduleFilename === __filename || moduleFilename.includes('iisnode')) {
  run();
}

export * from './src/main.server';

```
tsconfig.server.json

```
{
  "extends": "./tsconfig.app.json",
  "compilerOptions": {
    "outDir": "./out-tsc/app-server",
    "module": "commonjs",
    "types": [
      "node"
    ]
  },
  "files": [
    "src/main.server.ts",
    "server.ts"
  ],
  "angularCompilerOptions": {
    "entryModule": "./src/app/app.server.module#AppServerModule"
  }
}

```
angular.json
```
{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "coupons": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss"
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:browser",
          "options": {
            "outputPath": "dist/coupons/browser",
            "index": "src/index.html",
            "main": "src/main.ts",
            "polyfills": "src/polyfills.ts",
            "tsConfig": "tsconfig.app.json",
            "aot": true,
            "assets": ["src/favicon.ico", "src/assets"],
            "styles": [
              "node_modules/ngx-owl-carousel-o/lib/styles/prebuilt-themes/owl.carousel.min.css",
              "node_modules/ngx-owl-carousel-o/lib/styles/prebuilt-themes/owl.theme.default.min.css",
              "src/styles.scss"
            ],
            "scripts": []
          },
          "configurations": {
            "production": {
              "fileReplacements": [
                {
                  "replace": "src/environments/environment.ts",
                  "with": "src/environments/environment.prod.ts"
                }
              ],
              "optimization": true,
              "outputHashing": "all",
              "sourceMap": false,
              "extractCss": true,
              "namedChunks": false,
              "extractLicenses": true,
              "vendorChunk": false,
              "buildOptimizer": true,
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "2mb",
                  "maximumError": "5mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "6kb",
                  "maximumError": "10kb"
                }
              ]
            }
          }
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "options": {
            "browserTarget": "coupons:build"
          },
          "configurations": {
            "production": {
              "browserTarget": "coupons:build:production"
            }
          }
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "browserTarget": "coupons:build"
          }
        },
        "test": {
          "builder": "@angular-devkit/build-angular:karma",
          "options": {
            "main": "src/test.ts",
            "polyfills": "src/polyfills.ts",
            "tsConfig": "tsconfig.spec.json",
            "karmaConfig": "karma.conf.js",
            "assets": ["src/favicon.ico", "src/assets"],
            "styles": ["src/styles.scss"],
            "scripts": []
          }
        },
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": ["tsconfig.app.json", "tsconfig.spec.json", "e2e/tsconfig.json"],
            "exclude": ["**/node_modules/**"]
          }
        },
        "e2e": {
          "builder": "@angular-devkit/build-angular:protractor",
          "options": {
            "protractorConfig": "e2e/protractor.conf.js",
            "devServerTarget": "coupons:serve"
          },
          "configurations": {
            "production": {
              "devServerTarget": "coupons:serve:production"
            }
          }
        },
        "server": {
          "builder": "@angular-devkit/build-angular:server",
          "options": {
            "outputPath": "dist/coupons/server",
            "main": "server.ts",
            "tsConfig": "tsconfig.server.json"
          },
          "configurations": {
            "production": {
              "outputHashing": "media",
              "fileReplacements": [
                {
                  "replace": "src/environments/environment.ts",
                  "with": "src/environments/environment.prod.ts"
                }
              ],
              "sourceMap": false,
              "optimization": true
            }
          }
        },
        "serve-ssr": {
          "builder": "@nguniversal/builders:ssr-dev-server",
          "options": {
            "browserTarget": "coupons:build",
            "serverTarget": "coupons:server"
          },
          "configurations": {
            "production": {
              "browserTarget": "coupons:build:production",
              "serverTarget": "coupons:server:production"
            }
          }
        },
        "prerender": {
          "builder": "@nguniversal/builders:prerender",
          "options": {
            "browserTarget": "coupons:build:production",
            "serverTarget": "coupons:server:production",
            "routes": [
              "/"
            ]
          },
          "configurations": {
            "production": {}
          }
        }
      }
    }
  },
  "defaultProject": "coupons"
}

```

**Comment by neo** [comment]
This is probably b/c you haven't converted the app into a universal one; for instance, all the `window`, `document` and `Event` (in your case) usage needs to be wrapped into a "is in browser" check or adjusted in some other ways.

**Comment by alan-agius4** [comment]
Can you setup a minimal repro please?

You can read [here](https://github.com/angular/universal/blob/master/CONTRIBUTING.md#-submitting-an-issue) why this is needed. A good way to make a minimal repro is to create a new app via `ng new repro-app` and adding the minimum possible code to show the problem. Then you can push this repository to github and link it here.

**Comment by khaled6mmd** [comment]
got solved after updating ngx-owl-carousel-o: ^2.0.3 to 3.0.0
thanks guys
